[FLINK-2821] use custom Akka build to listen on all interfaces

mxm opened the pull request:
This uses Flakka (a custom Akka 2.3 build) to resolve the issue that
the bind address needs to be matching the external address of the
JobManager. With the changes applied, we can now bind to all
interfaces, e.g. via 0.0.0.0 (IPv4) or :: (Ipv6).

For this to work properly, the configuration entry
JOB_MANAGER_IPC_ADDRESS now represents the external address of the
JobManager. Consequently, it should not be resolved to an IP address
anymore because it may not be resolvable from within containered
environments. Akka treats this address as the logical address. Any
messages which are not tagged with this address will be received by
the Actor System (because we listen on all interfaces) but will be
dropped subsequently. In addition, we need the external address for
the JobManager to be able to publish its address to Zookeeper for HA
setups.

Flakka: https://github.com/mxm/flakka 
Main patch applied: akka/akka#15610

StephanEwen commented on flink-core/src/main/java/org/apache/flink/util/NetUtils.java:
```diff
@@ -309,6 +313,19 @@ public static ServerSocket createSocketFromPorts(Iterator<Integer> portsIterator
+	 * Returns the wildcard address to listen on all interfaces.
+	 * @return Either 0.0.0.0 or :: depending on the IP setup.
+	 */
+	public static String getWildcardIPAddress() {
```
You could move that code into a static initializer. That way you don't need to lock and check if it was set before.

mxm replied:
I chose not to do that to avoid throwing exceptions in a static initializer. The current implementation preserves the context of the exception. I'm ok with initializing it statically.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1891,8 +1891,8 @@ object JobManager {
          executionMode: JobManagerMode,
-         listeningHost: String,
-         listeningPortRange: java.util.Iterator[Integer]) =
+         reachableHost: String,
```
What does "reachableHost" mean here? I understand that it is not really the "listeningHost" any more (that is the wildcard address), but I don't understand what "reachableHost" should mean.

mxm replied:
It is the consistent external hostname of the JobManager.

StephanEwen replied:
How about calling it simply `hostname` or `externalHostname` then?

mxm replied:
Agreed. I've switched terms during some iterations on the changes. I think `externalHostname` is the better alternative.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -2729,11 +2727,11 @@ object JobManager {
     val (protocol, hostname, port) = TaskManager.getAndCheckJobManagerAddress(config)
 
-    var hostPort: InetSocketAddress = null
+    var hostPort: InetSocketAddress = new InetSocketAddress(hostname, port)
```
Does this change alter anything compared to before? At a first glance, both construct the same `InetSocketAddress` and both needs to hostname to be resolvable.
Is it important that the `InetSocketAddress` contains the hostname, rather than the IP address?

mxm replied:
It is just a check to see if we expect messages from a resolvable hostname. It is not necessary to do that but I found test cases rely on that. Also, it improves error reporting for the user in case a non-resolvable hostname was chosen.

StephanEwen commented:
It's a cool fix to help Flink run in container environments.

There is one concern I have with the code, though: Previously, we decided explicitly to encode IP addresses in the Akka URLs, because Akka requires an exact match of the host in the URL.

That has two problems:

  - All nodes need to agree to use either hostname or address. Enforcing to use the address was possible, enforcing to use the hostname not.

  - More importantly: IPv6 has multiple valid representations of the same address. If different nodes have the same address but represent it as a string differently, the messages would not arrive. That's why we have a "normalizing" encoding step that makes sure all URLs would be the same. That step relied on translating 

I see two dangers now that may occur:
  - If different parts of the code or the JDK do a subtle change of behavior (i.e. resolve the InetAddress), then some nodes may have a hostname in the URL, others an address.

  - If two machines have a slightly different network configuration (especially concerning preferences to represent/encode IPv6 addresses) and the users set the IP address as the JobManager host, then they might create different Akka URLs and the machines cannot talk to each other again.

I am a bit unsure how to proceed from here. Is there any way we can keep using IP addresses in the Akka URLs? Or does that just inherently not make sense with "dynamic hostnames" as they are used in container environments?

mxm commented:
Thanks for checking out the code!

>If different parts of the code or the JDK do a subtle change of behavior (i.e. resolve the InetAddress), then some nodes may have a hostname in the URL, others an address.

Fair point. Let's remove InetAddress and use a String instead.

>If two machines have a slightly different network configuration (especially concerning preferences to represent/encode IPv6 addresses) and the users set the IP address as the JobManager host, then they might create different Akka URLs and the machines cannot talk to each other again.

A problem of the old method was that hostnames might resolve differently depending on the container context. So that method was not reliable either. Using IP addresses will work with the new method. They just have to be consistent on all node configurations. I don't think that is a problem because a configuration is usually created once and then copied over to all nodes.

>I am a bit unsure how to proceed from here. Is there any way we can keep using IP addresses in the Akka URLs? Or does that just inherently not make sense with "dynamic hostnames" as they are used in container environments?

For this PR to address the core problems of FLINK-2821, we have to avoid resolving the hostname because the IP address would represent the internal container address which may be unreachable from the outside; even if it were resolvable, Akka would drop the messages because of its exact URL match policy. Using the hostname which now acts as a purely logical address ultimately is a more reliable way across different network environments. We just have to make sure we document the new requirement that the JobManager address (JOB_MANAGER_IPC_ADDRESS) is consistent across all Flink cluster nodes.

StephanEwen commented:
I have to dig back a bit (long time since I worked on that), but I thought the IPv6 addresses issue was quite tricky one and the normalization code for address representations was necessary.

Can we try and preserve both in some way? When a logical hostname is specified, we use that one in the Akka URLs. When an IP address is specified, we normalize it and use it.

That way we only "loose" the ability to use mixed hostname/IP-address configurations, but that would be okay, I guess.

mxm commented:
>Can we try and preserve both in some way? When a logical hostname is specified, we use that one in the Akka URLs. When an IP address is specified, we normalize it and use it.

Detecting whether we have an IPv4/IPv6 address or a hostname should be possible. Let me see if I can incorporate that in the PR.

mxm commented:
@StephanEwen I've updated the pull request to incorporate your suggestions. When an IPv6 address is specified, we format it like in the current code base. When a hostname is specified, we do some simple validation but do not resolve it. IPv4 addresses are simply used as-is. 

I hope that this PR becomes more mergeable in this state. With the help of some eager users we can further test this in a cluster environment.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala:
```diff
@@ -102,21 +102,24 @@ object AkkaUtils {
-        val remoteConfig = getRemoteAkkaConfig(configuration, hostString, port)
+
+        val remoteConfig = getRemoteAkkaConfig(configuration,
+          NetUtils.getWildcardIPAddress, port,
```
Maybe we could add a comment here that we choose the wildcard ip to bind to all network interfaces.

mxm replied:
+1

StephanEwen commented on flink-runtime/pom.xml:
```diff
@@ -193,8 +193,8 @@ under the License.
-			<groupId>com.typesafe.akka</groupId>
-			<artifactId>akka-testkit_${scala.binary.version}</artifactId>
+			<groupId>com.data-artisans</groupId>
+			<artifactId>flakka-testkit_${scala.binary.version}</artifactId>
```
I think this dependency be in `test` scope.

mxm replied:
This probably needs to be changed independently of this PR.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala:
```diff
@@ -216,12 +219,19 @@ object AkkaUtils {
    */
   private def getRemoteAkkaConfig(configuration: Configuration,
-                                  hostname: String, port: Int): Config = {
+                                  hostname: String, port: Int,
```
Would be good to call `hostname` here `bind address`.

mxm replied:
+1

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/util/NetUtils.java:
```diff
@@ -111,7 +115,51 @@ public static int getAvailablePort() {
+			// We don't allow these in hostnames
+			Preconditions.checkArgument(!host.startsWith("."));
+			Preconditions.checkArgument(!host.endsWith("."));
+			Preconditions.checkArgument(!host.contains(":"));
```
Maybe we could add a clarifying error message here for the user.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala:
```diff
@@ -216,12 +219,19 @@ object AkkaUtils {
+                                  hostname: String, port: Int,
+                                  externalHostname: String, externalPort: Int): Config = {
+
+    LOG.info(s"Using binding address $hostname:$port" +
```
Not too bad here, but in general, I think having loggers in util classes is not a very good idea. It typically leads to the log lines showing up very often and at strange places. The information should usually be logged by the caller of the functions.

mxm replied:
That's right. Removing this statement because this is also logged at JobManager startup.

mxm commented:
Thanks you @tillrohrmann  and @StephanEwen. I've addressed your comments. I'll have to redeploy Flakka because the staging repository which this PR used, has been dropped in the meantime. I will update the PR tomorrow to use the Maven central servers.

mxm commented:
I've added the new staging repository to test the PR changes. Also, the repository is currently deploying to Maven central.

StephanEwen commented:
Looks good to me now, +1 to merge this

mxm commented:
Thank you, will go ahead and merge then.

schmichri commented:
when is this planned to be released?

StephanEwen commented:
There is a release candidate out: https://github.com/apache/flink/tree/release-1.2.0-rc0
Download at http://people.apache.org/~rmetzger/flink-1.2.0-rc0/

The official release comes when the testing phase is done, which should be soon.

If you can try the release candidate and tell us if it worked for your scenario, that would help with the release testing.

schmichri commented:
short: it works!

long: my scenario a docker-compose setup: 

`  jobmanager:
    image: flink:1.2.0_rc0
    container_name: "jobmanager"
    expose:
      - "6123"
      - "8081"
    ports:
      - "8081:8081"
      - "10.0.0.11:6123:6123"
    command: jobmanager
    environment:
     - JOB_MANAGER_RPC_ADDRESS=10.0.0.11
`
and the taskmanager(s):
`  taskmanager:
    image: flink:1.2.0_rc0
    expose:
      - "6121"
      - "6122"
    command: taskmanager
      - "jobmanager:jobmanager"
    environment:
      - JOB_MANAGER_RPC_ADDRESS=10.0.0.11
    extra_hosts:
      - "jobmanager:10.0.0.11"`

1. the jobmanager webinterface works now
2. the taskmanager can connect now to the jobmanager 
previous errormessage: 
> dropping message [class akka.actor.ActorSelectionMessage] for non-local recipient [Actor[akka.tcp://flink@10.0.0.11:6123/]] arriving at [akka.tcp://flink@10.0.0.11:6123] inbound addresses are [akka.tcp://flink@172.31.0.2:6123]
> flink-master    | 2017-01-09 15:30:39,865 ERROR akka.remote.EndpointWriter

thats very cool!

peay commented:
This did not work for me with `1.2.0-rc2`. Is there something special to do to make it bind on `0.0.0.0` now?

I am still getting:

```
dropping message [class akka.actor.ActorSelectionMessage] for non-local recipient [Actor[akka.tcp://flink@10.50.50.167:6123/]] arriving at [akka.tcp://flink@10.50.50.167:6123] inbound addresses are [akka.tcp://flink@jobmanager.flink.svc.cluster.local:6123]
```

with `JOB_MANAGER_RPC_ADDRESS=jobmanager.flink.svc.cluster.local`.

mxm commented:
>edit: Re-reading the PR description, I actually got it to work by setting jobmanager.rpc.address to the external IP and getting rid of the hostname.

Glad you were able to solve the problem! :) Akka requires that all messages are tagged with the same address which was used during initialization of the receiver actor system. While this PR gets rid of the issue that an address might not be available for binding during initialization of the receiver, it still requires you to use a consistent address.
